subsys/bluetooth/host/att.c

```diff
@@ -176,35 +175,57 @@ static struct bt_att_req cancel;
  */
 static k_tid_t att_handle_rsp_thread;
 
-#define bt_att_tx_meta_data(buf) (((struct bt_att_tx_meta *)net_buf_user_data(buf))->data)
+static struct bt_att_tx_meta_data tx_meta_data_storage[CONFIG_BT_ATT_TX_COUNT];
 
-static struct bt_att_tx_meta_data tx_meta_data_storage[CONFIG_BT_CONN_TX_MAX];
-K_FIFO_DEFINE(free_att_tx_meta_data);
+struct bt_att_tx_meta_data *bt_att_get_tx_meta_data(const struct net_buf *buf);
+static void att_on_sent_cb(struct bt_att_tx_meta_data *meta);
 
-static struct bt_att_tx_meta_data *tx_meta_data_alloc(k_timeout_t timeout)
+static void att_tx_destroy(struct net_buf *buf)
 {
-	/* The meta data always get freed in the system workqueue,
-	 * so if we're in the same workqueue but there are no immediate
-	 * contexts available, there's no chance we'll get one by waiting.
+	struct bt_att_tx_meta_data *p_meta = bt_att_get_tx_meta_data(buf);
+	struct bt_att_tx_meta_data meta;
+
+	LOG_DBG("destroy %p", buf);
+
+	/* Destroy the buffer first, as the callback may attempt to allocate a
+	 * new one for another operation.
 	 */
-	if (k_current_get() == &k_sys_work_q.thread) {
-		return k_fifo_get(&free_att_tx_meta_data, K_NO_WAIT);
+	meta = *p_meta;
+
+	if (IS_ENABLED(CONFIG_BT_TESTING)) {
+		/* Still do it in test code */
+		memset(p_meta, 0xFF, sizeof(*p_meta));
```

Comment:
What is the purpose of setting it to `0xFF` instead of 0 by default? 

We want to reset it, right? And the metadata are `0` by default since they are `static`. If we want to reset it to `0xFF`, then we should also initialize all the metadata at boot to `0xFF`, else we should just use `0` here.

Reply:
it's to catch misuses in the time between a destroy (init to 0xff) and a re-allocation (init to 0).
If that sounds stupid, I'll just remove this line.